### PR TITLE
Move `DifficultyMultiplierDisplay`'s layout and colour logic to a base class

### PR DESCRIPTION
**Pull request by Feodor0090**
This PR adds an abstract class for mods' effect displays with layout and colour logic to share them between `DifficultyMultiplierDisplay`, `BpmDisplay` and others those will be possibly added in future.

### CODE REVIEW

File: osu.Game/Overlays/Mods/ModsEffectDisplay.cs [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+                        Width = value_area_width + ModSelectPanel.CORNER_RADIUS
+                    },
+                    new GridContainer
+                    {
+                        RelativeSizeAxes = Axes.Y,
+                        AutoSizeAxes = Axes.X,
+                        ColumnDimensions = new[]
+                        {
+                            new Dimension(GridSizeMode.AutoSize),
+                            new Dimension(GridSizeMode.Absolute, value_area_width)
```
Comment (bdach) [inline osu.Game/Overlays/Mods/ModsEffectDisplay.cs:71]: Not sure this fixed width spec is going to work well in all cases.

File: osu.Game/Overlays/Mods/ModsEffectDisplay.cs [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        {
+            labelBackground.Colour = colourProvider.Background4;
+        }
+
+        /// <summary>
+        /// Fades colours of text and its background according to displayed value.
+        /// </summary>
+        /// <param name="difference">Difference between actual value and default value.
+        /// Negative leads to green color, positive to red. Can be obtained via CompareTo(defaultValue).</param>
+        protected void SetColours(int difference)
```
Comment (bdach) [inline osu.Game/Overlays/Mods/ModsEffectDisplay.cs:125]: Not a huge fan of this API design. I see what you're going for - implementations of this component may have different ideas whether a larger value of their respective metric is, depending on whether it makes gameplay easier or harder - but as is, you have to remember to call `SetColours(0)` manually to get the component into a sane initial state.

A better route may look something like (loose sketch):

```csharp
public class ModsEffectDisplay<TValue> : IHasCurrentValue<TValue>
{
    public Bindable<TValue> Value { get; }

    public enum ModEffect
    {
         NoChange,
         DifficultyReduction,
         DifficultyIncrease
    }

    public abstract ModEffect CalculateEffectFor(TValue value);

    protected override LoadComplete()
    {
        Current.BindValueChanged(current =>
        {
            var effect = CalculateModEffect(current.NewValue);
            setColours(effect);
        }
    }
}
```

Gains from this design:

- Implementations get the `Current` bindable for free
- Implementations don't have to manually manage the colours - all they need to do is implement the pure abstract method that determines what the new value represents difficulty-wise

**Comment by bdach** [comment]
Also, this PR only makes sense insofar as https://github.com/ppy/osu/pull/19929 makes sense, which I don't know if it does (the [design issues](https://github.com/ppy/osu/pull/19929#issuecomment-1225212075) are still largely unaddressed).

**Comment by peppy** [comment]
I'm so confused by this direction. Can we remove the generics for a start? Is there a reason this needs to be `abstract` in the first place? Like can't all the counter logic be in the single base class and just have the implementation decide whether it wants to show the "x" icon and call it a day?

**Comment by bdach** [comment]
The difficulty multiplier box goes green for difficulty below 1.0x and red for difficulty above 1.0x. As long as this invariant holds for every future usage of this box (that lower value = easier, higher value = harder) holds, you could simplify a lot of this, yes. But I don't know that it does hold. Duration may be one case where it wouldn't.

**Comment by peppy** [comment]
I think it's fine to leave that part is `virtual`, but I'm not sure about the counter logic being done in each implementation. Dunno, just structurally feels a bit weird (like the `abstract` class just has helper logic which you're required to know how to use?).

**Review by smoogipoo — APPROVED** [review]

File: osu.Game/Overlays/Mods/ModsEffectDisplay.cs [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+        /// </summary>
+        /// <param name="value">Value to calculate for. Will arrive from <see cref="Current"/> bindable once it changes.</param>
+        /// <returns>Effect of the value.</returns>
+        /// <remarks>
+        /// Default implementation compares the value with <see cref="Current"/>.<see cref="Bindable{T}.Default"/>, bigger value counts as difficulty increase.
+        /// </remarks>
+        protected virtual ModEffect CalculateEffect(double value)
+        {
+            return CalculateForSign(value.CompareTo(Current.Default));
+        }
```
Comment (smoogipoo) [inline osu.Game/Overlays/Mods/ModsEffectDisplay.cs:63]: This flow is kinda weird.
1. Why is `CalculateForSign` not inlined in here?
2. How is a consumer supposed to implement this? Override `CalculateEffect` and compare it themselves?

I feel a better implementation is to remove this method, inline it in the single usage site, rename `CalculateForSign` to `GetEffectForComparison`, and make that method virtual.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Overlays/Mods/ModsEffectDisplay.cs (hunk 3 above)
Comment (Feodor0090) [reply osu.Game/Overlays/Mods/ModsEffectDisplay.cs:63]: > How is a consumer supposed to implement this? Override CalculateEffect and compare it themselves?

Initially i thought that yes. But now i understand that to flip effect consumer can override `CalculateForSign` like `return base.CalculateForSign(-comparison);`. So i agree with your point.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**